osquery/filesystem/filesystem.cpp

```diff
@@ -147,6 +148,120 @@ Status listFilesInDirectory(const boost::filesystem::path& path,
   }
 }
 
+Status listDirectoriesInDirectory(const boost::filesystem::path& path,
+                            std::vector<std::string>& results) {
+  try {
+    if (!boost::filesystem::exists(path)) {
+      return Status(1, "Directory not found");
+    }
+
+    if (!boost::filesystem::is_directory(path)) {
+      return Status(1, "Supplied path is not a directory");
+    }
+
+    boost::filesystem::directory_iterator begin_iter(path);
+    boost::filesystem::directory_iterator end_iter;
+    for (; begin_iter != end_iter; begin_iter++) {
+      if (boost::filesystem::is_directory(begin_iter->path())) {
+        results.push_back(begin_iter->path().string());
+      }
+    }
+
+    return Status(0, "OK");
+  } catch (const boost::filesystem::filesystem_error& e) {
+    return Status(1, e.what());
+  }
+}
+
+Status doFileListing(const boost::filesystem::path& fs_path, std::vector<std::string>& results, std::vector<std::string>& components){
+  if(components[components.size()-1][0] == '*' && components[components.size()-1].length() == 1){
+    std::vector<std::string> files;
+    Status stat = listFilesInDirectory(fs_path.parent_path(), files);
+    if(stat.getCode() != 0){return stat;}
+    for(auto& file : files){
+        results.push_back(file);
+    }
+    return Status(0, "OK");
+  }
+  //If we get here, that means we've reached the last component of the original path.
+  //Check to make sure the path we've arrived at is valid at all.
+  try{
+    if(!boost::filesystem::exists(fs_path)){return Status(0, "OK");}
+  }catch(boost::filesystem::filesystem_error error){
+    VLOG(1) << "Permission was denied to access: " << fs_path.string();
+    return Status(0, "OK");
+  }
+  if(boost::filesystem::is_directory(fs_path)){
+    std::vector<std::string> files;
+    Status stat = listFilesInDirectory(fs_path, files);
+    if(stat.getCode() != 0){return stat;}
```

Comment:
Also, never put entire if statements on one line, always break it up onto three:

```cpp
if (stat.ok()) {
  return stat;
}
```